### PR TITLE
Affordable rentals

### LINKED ISSUE #74: Normalize addresses
Addresses are in a variety of formats, so we should normalize them to one standard string. This will aid connecting records without a proper FK.

### PR DESCRIPTION
**Pull request by sbuss**
Replaces #75, Fixes #74 and #52

### CODE REVIEW

File: schemaless/sources.py [hunk 1]
```diff
@@ -50,6 +55,57 @@ def get_value_str(self, record):
...
+                    print("WARN2: Unable to format address %s" % addr)
+                    return ""
+
+            if not addr['city']:
+                addr['city'] = "San Francisco"
+            if not addr['state']:
+                addr['state'] = "California"
+
+        try:
+            addr = normalize_address_record(addr)
```
Comment (ilakyapal) [inline schemaless/sources.py]: Why do we normalize twice again?

Comment (ilakyapal) [reply schemaless/sources.py]: To make sure everything is normalized after filling in any gaps for zip/city/state right?

File: schemaless/sources.py [hunk 2]
```diff
@@ -70,9 +126,26 @@ def yield_records(self):
...
+                        continue
+
+                    if line_key not in line:
+                        continue
+
+                    val = (line.get(line_key) or "").strip()
+                    if val:
+                        ret[ret_key] = val
+
+                for key, field in calc_fields.items():
```
Comment (ilakyapal) [inline schemaless/sources.py]: This means that the normalized addresses gets written to the schemaless file right? Did we want to start including fields we've done our own processing on in the schemaless file (thought we weren't doing that)?

File: gen-test-data.sh [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
-xzcat data/ppts/2019-06-02-ppts.csv.xz | egrep -i "(1950 mission)|(8 10th)|(429 beale)|(430 main)|(2015-011205)|(201404304554)|(201705237369)|(2015-014058PRJ)" >> testdata/ppts-two.csv
+xzcat data/ppts/2019-06-02-ppts.csv.xz | egrep -i "(1950 mission)|(8 10th)|(429 beale)|(430 main)|(2015-011205)|(201404304554)|(201705237369)|(2015-014058PRJ)|((340 Valencia)|(2002.0809))|(2451 Sacramento)" >> testdata/ppts-two.csv
 
 head -n1 data/mohcd/mohcd-pipeline-2020-01-30.csv > testdata/mohcd-pipeline.csv
 cat data/mohcd/mohcd-pipeline-2020-01-30.csv | egrep -i "(1950 mission)|(2015-011205)|(transbay)|(2011-005)|(2016-023)|(2015-014058PRJ)" >> testdata/mohcd-pipeline.csv
 
-xzcat data/pts/2020-01-24-pts-after-2013.csv.xz | head -n1 > testdata/pts.csv
-xzcat data/pts/2020-01-24-pts-after-2013.csv.xz | egrep -i "(201705318009)|(201706058373)|(201711284964)|(201801118389)|(201404304554)|(201705237369)" >> testdata/pts.csv
+head -n1 data/mohcd/bmr-2019-09-06.csv > testdata/bmr.csv
+cat data/mohcd/bmr-2019-09-06.csv | egrep -i "(Valencia Gardens)|(Kennedy Towers)" >> testdata/bmr.csv
```
[the remark this answers is not in the document]
Comment (sbuss) [reply gen-test-data.sh]: Gotcha. Renamed to `AffordableRentalPortfolio`

File: schemaless/sources.py (hunk 1 above)
Comment (sbuss) [reply schemaless/sources.py]: There are a lot of addresses that look like "123 Main St 94102". This gets parsed into a dict that looks like:

```python
{'address_line_1': '123 MAIN ST',
 'address_line_2': None,
 'city': None,
 'state': None,
 'postal_code': '94102'}
```

We want `city` and `state` to be populated, too, so I add them in and then normalize once more to ensure it's still valid.

File: schemaless/sources.py (hunk 2 above)
Comment (sbuss) [reply schemaless/sources.py]: Yep, it gets written out.

> Did we want to start including fields we've done our own processing on in the schemaless file (thought we weren't doing that)?

I didn't think about that. I guess there's no reason to write it out, but is there a strong reason _not_ to?

File: schemaless/sources.py [hunk 4]
```diff
@@ -218,10 +303,12 @@ class TCO(Source):
...
         'Building Address': 'address',
+        'address_norm': Address('address'),
         'Date Issued': 'date_issued',
         'Document Type': 'building_permit_type',
         'Number of Units Certified': 'num_units',
     }
+    DATA_SF = "https://data.sfgov.org/Housing-and-Buildings/Dwelling-Unit-Completion-Counts-by-Building-Permit/j67f-aayr"  # NOQA
 
 
 class MOHCDInclusionary(Source):
```
[the remark this answers is not in the document]
Comment (sbuss) [reply schemaless/sources.py]: done

File: schemaless/sources.py (hunk 2 above)
Comment (ilakyapal) [reply schemaless/sources.py]: Approving, as discussed offline, we just won't write the normalized address to the schemelass for now

Comment (sbuss) [reply schemaless/sources.py]: Changed to not write out computed fields in 627caddee44cd473e810eaeb41fe9eeb8099b8b5

**Pull request merged**